### ISSUE TITLE
Custom metrics don't seem to work with parallel processing

### ISSUE THREAD

**Issue by py9mrg** [tidymodels/tune#479]
Hello,

I don't *think* this is related to other issues but could be wrong, so am posting a new issue. I don't seem to be able to get cutom metrics to work with parallel processing:

``` r
library(tidymodels)

rf_mod <- 
  rand_forest() %>% 
  set_engine("ranger") %>% 
  set_mode("regression")

set.seed(345)
folds <- vfold_cv(mtcars, v = 10)

rf_wf <- 
  workflow() %>%
  add_model(rf_mod) %>%
  add_formula(mpg ~ .)

### works fine with default metric
set.seed(456)
rf_fit_rs <- 
  rf_wf %>% 
  fit_resamples(folds)

### create custom metric
rmse_trans_vec <- function(truth, estimate, na_rm = TRUE, ...) {

  rmse_trans_impl <- function(truth, estimate) {
    estimate <- 2 * estimate
    sqrt(mean((truth - estimate) ^ 2))
  }

  metric_vec_template(
    metric_impl = rmse_trans_impl,
    truth = truth,
    estimate = estimate,
    na_rm = na_rm,
    cls = "numeric",
    ...
  )

}

rmse_trans <- function(data, ...) {
  UseMethod("rmse_trans")
}

rmse_trans <- new_numeric_metric(rmse_trans, direction = "minimize")

rmse_trans.data.frame <- function(data, truth, estimate, na_rm = TRUE, ...) {

  metric_summarizer(
    metric_nm = "rmse_trans",
    metric_fn = rmse_trans_vec,
    data = data,
    truth = !! enquo(truth),
    estimate = !! enquo(estimate),
    na_rm = na_rm,
    ...
  )

}

### works fine sequentially
set.seed(456)
rf_fit_rs <- 
  rf_wf %>% 
  fit_resamples(
      folds,
      metrics = metric_set(rmse_trans)
    )

### doesn't work in parallel (running on Windows)
library(doFuture)
#> Loading required package: foreach
#> 
#> Attaching package: 'foreach'
#> The following objects are masked from 'package:purrr':
#> 
#>     accumulate, when
#> Loading required package: future
registerDoFuture()
plan(multisession, workers = 2) # neither multisession nor cluster work

set.seed(456)
rf_fit_rs <- 
  rf_wf %>% 
  fit_resamples(
      folds,
      metrics = metric_set(rmse_trans)
    )
#> x Fold01: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold02: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold03: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold04: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold05: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold06: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold07: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold08: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold09: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> x Fold10: internal: Error:
#> ! In metric: `rmse_trans`
#> no applicable method for 'rms...
#> Warning: All models failed. See the `.notes` column.

### again, works sequentially
plan(sequential)

set.seed(456)
rf_fit_rs <- 
  rf_wf %>% 
  fit_resamples(
      folds,
      metrics = metric_set(rmse_trans)
    )
```

<sup>Created on 2022-03-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by juliasilge** [comment]
Oooooo, this is challenging. We [explain a bit here](https://www.tmwr.org/grid-search.html#access-to-global-variables) about how different parallelization technologies may not have access to the global environment. Making a custom metric is more like making a [custom parsnip model](https://www.tidymodels.org/learn/develop/models/) or a [custom recipe step](https://www.tidymodels.org/learn/develop/recipes/) than passing in a variable from the global environment, so you can't just `!!` your way to success. 

In situations like these, one option is to make a little package for the custom object and then send that package to the parallel workers. (Notice [the `pkgs` argument](https://tune.tidymodels.org/reference/control_grid.html) for a tuning control object.)

Another option is to switch to a parallel technology that passes the whole global environment to the workers, like `multicore`. (Does `multicore` work on Windows? I think so?)

``` r
library(tidymodels)

rf_mod <- 
    rand_forest() %>% 
    set_engine("ranger") %>% 
    set_mode("regression")

set.seed(345)
folds <- vfold_cv(mtcars, v = 10)

rf_wf <- 
    workflow() %>%
    add_model(rf_mod) %>%
    add_formula(mpg ~ .)

### create custom metric
rmse_trans_vec <- function(truth, estimate, na_rm = TRUE, ...) {
    
    rmse_trans_impl <- function(truth, estimate) {
        estimate <- 2 * estimate
        sqrt(mean((truth - estimate) ^ 2))
    }
    
    metric_vec_template(
        metric_impl = rmse_trans_impl,
        truth = truth,
        estimate = estimate,
        na_rm = na_rm,
        cls = "numeric",
        ...
    )
    
}

rmse_trans <- function(data, ...) {
    UseMethod("rmse_trans")
}

rmse_trans <- new_numeric_metric(rmse_trans, direction = "minimize")

rmse_trans.data.frame <- function(data, truth, estimate, na_rm = TRUE, ...) {
    
    metric_summarizer(
        metric_nm = "rmse_trans",
        metric_fn = rmse_trans_vec,
        data = data,
        truth = !! enquo(truth),
        estimate = !! enquo(estimate),
        na_rm = na_rm,
        ...
    )
    
}

library(doFuture)
#> Loading required package: foreach
#> 
#> Attaching package: 'foreach'
#> The following objects are masked from 'package:purrr':
#> 
#>     accumulate, when
#> Loading required package: future
registerDoFuture()
plan(multicore, workers = 2) # neither multisession nor cluster work
#> Warning in supportsMulticoreAndRStudio(...): [ONE-TIME WARNING] Forked
#> processing ('multicore') is not supported when running R from RStudio
#> because it is considered unstable. For more details, how to control forked
#> processing or not, and how to silence this warning in future R sessions, see ?
#> parallelly::supportsMulticore

set.seed(456)
rf_wf %>% 
    fit_resamples(
        folds,
        metrics = metric_set(rmse_trans)
    )
#> # Resampling results
#> # 10-fold cross-validation 
#> # A tibble: 10 × 4
#>    splits         id     .metrics         .notes          
#>    <list>         <chr>  <list>           <list>          
#>  1 <split [28/4]> Fold01 <tibble [1 × 4]> <tibble [0 × 3]>
#>  2 <split [28/4]> Fold02 <tibble [1 × 4]> <tibble [0 × 3]>
#>  3 <split [29/3]> Fold03 <tibble [1 × 4]> <tibble [0 × 3]>
#>  4 <split [29/3]> Fold04 <tibble [1 × 4]> <tibble [0 × 3]>
#>  5 <split [29/3]> Fold05 <tibble [1 × 4]> <tibble [0 × 3]>
#>  6 <split [29/3]> Fold06 <tibble [1 × 4]> <tibble [0 × 3]>
#>  7 <split [29/3]> Fold07 <tibble [1 × 4]> <tibble [0 × 3]>
#>  8 <split [29/3]> Fold08 <tibble [1 × 4]> <tibble [0 × 3]>
#>  9 <split [29/3]> Fold09 <tibble [1 × 4]> <tibble [0 × 3]>
#> 10 <split [29/3]> Fold10 <tibble [1 × 4]> <tibble [0 × 3]>
```

<sup>Created on 2022-04-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

You'll notice the warning here from doFuture about using multicore.
